x-pack/plugin/profiling/src/main/java/org/elasticsearch/xpack/profiling/TransportGetFlamegraphAction.java

```diff
@@ -0,0 +1,287 @@
+import org.elasticsearch.core.SuppressForbidden;
+import org.elasticsearch.tasks.Task;
+import org.elasticsearch.transport.TransportService;
+
+import java.io.File;
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.SortedMap;
+import java.util.TreeMap;
+
+public class TransportGetFlamegraphAction extends HandledTransportAction<GetStackTracesRequest, GetFlamegraphResponse> {
+    private static final Logger log = LogManager.getLogger(TransportGetFlamegraphAction.class);
+    private static final StackFrame EMPTY_STACKFRAME = new StackFrame("", "", 0, 0);
+
+    private final NodeClient nodeClient;
+    private final TransportService transportService;
+
+    @Inject
+    public TransportGetFlamegraphAction(NodeClient nodeClient, TransportService transportService, ActionFilters actionFilters) {
+        super(GetFlamegraphAction.NAME, transportService, actionFilters, GetStackTracesRequest::new);
+        this.nodeClient = nodeClient;
+        this.transportService = transportService;
+    }
+
+    @Override
+    protected void doExecute(Task task, GetStackTracesRequest request, ActionListener<GetFlamegraphResponse> listener) {
+        Client client = new ParentTaskAssigningClient(this.nodeClient, transportService.getLocalNode(), task);
+        long start = System.nanoTime();
+        client.execute(GetStackTracesAction.INSTANCE, request, new ActionListener<>() {
+            @Override
+            public void onResponse(GetStackTracesResponse response) {
+                long responseStart = System.nanoTime();
+                try {
+                    GetFlamegraphResponse flamegraphResponse = buildFlamegraph(response);
+                    log.debug(
+                        "getFlamegraphAction took ["
+                            + (System.nanoTime() - start) / 1_000_000.0d
+                            + "] ms (processing response: ["
+                            + (System.nanoTime() - responseStart) / 1_000_000.0d
+                            + "] ms."
+                    );
+                    listener.onResponse(flamegraphResponse);
+                } catch (Exception ex) {
+                    listener.onFailure(ex);
+                }
+            }
+
+            @Override
+            public void onFailure(Exception e) {
+                listener.onFailure(e);
+            }
+        });
+    }
+
+    static GetFlamegraphResponse buildFlamegraph(GetStackTracesResponse response) {
+        // TODO: Are full seconds good enough? (they probably are)
+        long totalSeconds = 0;
+        if (response.getStartTime() != null && response.getEndTime() != null) {
+            Duration observedDuration = Duration.between(response.getStartTime(), response.getEndTime());
+            totalSeconds = (observedDuration.getNano() > 0) ? observedDuration.getSeconds() + 1 : observedDuration.getSeconds();
+        }
```

Comment:
Is the `observedDuration` equal to the duration between the first and last returned samples in the response? For instance, if we request 24 hours of data (86400 seconds) but we only have 1 hour of data (3600 seconds), then the `observedDuration` is 3600.

I think this is the case based on my local testing. This will be a problem in Kibana because differential flamegraphs relies on `TotalSeconds` being equal to the duration in the original request when Kibana scales the samples in the two flamegraphs based on time.

Reply:
Thanks for the context. You're right indeed: the duration is based on the first and last observed sample that match the provided filter. If Kibana still needs to do scaling / normalization I wonder whether it makes sense to pass that additional context around within Kibana? In that case I'd tend to exclude this property at all from the response or only keep it for diagnostic purposes. Wdyt?

Reply:
I agree that it makes sense to remove this property since it is redundant at this time. It was originally returned in the first implementation. It was then carried over to Kibana (where it was useful for testing), but it's no longer needed and adds unnecessary work.

Reply:
Great. I've removed `TotalSeconds` from the response in https://github.com/elastic/elasticsearch/pull/99091/commits/98175d0dd35fa813193f71aa85ccde28ad7019bb.